src/computerAlgorithms.js

```diff
@@ -0,0 +1,202 @@
+      let [isZeroOnAngle,] = findMarkOn(squares, 'O', 'angle');
+
+      if (isZeroOnAngle) {
+        return getRandomAnglePosition(squares);
+      } else {
+        return getAnglePositionOnFriendlyLine(squares);
+      }
+    } else {
+      for(let i = 0; i < diagonalLines.length; i ++) {
+        const [a, , c] = diagonalLines[i];
+
+        if (squares[a] === 'X' && !squares[c]) {
+          return c;
+        } else if (squares[c] === 'X' && !squares[a]) {
+          return a;
+        }
+      }
+    }
+  }
+
+  if (squares[4] === 'X') {
+    return getRandomAnglePosition(squares);
+  }
+}
+
+function findMarkOn(squares, mark, where) {
+  let i = where === 'angle' ? 0 : 1;
+  let isMarkOn = false;
+  let markPosition = null;
+
+  while(i < squares.length) {
+    if (squares[i] === mark && i !== 4) {
+      isMarkOn = true;
+      markPosition = i;
+      break;
+    }
+    i += 2;
+  }
+
+  return [isMarkOn, markPosition];
+}
+
+function findRepetitionsAlgorithm(squares, playerGoesFirst){
+  const computerSymbol = playerGoesFirst ? 'O' : 'X';
+  const playerSymbol = computerSymbol === 'X' ? 'O' : 'X';
+
+  if (findRepetitions(squares, playerSymbol) !== null) {
+    return findRepetitions(squares, playerSymbol);
+  } else if (findRepetitions(squares, computerSymbol) !== null) {
+    return findRepetitions(squares, computerSymbol);
+  }
+
+  return null;
+}
+
+function findRepetitions(squares, symbol){
+  for (let i = 0; i < lines.length; i++) {
+    const [a, b, c] = lines[i];
+
+    if (squares[a] === symbol || squares[b] === symbol || squares[c] === symbol) continue;
+
+    if (squares[a] && squares[a] === squares[b]) {
+      return c;
+    } else if (squares[b] && squares[b] === squares[c]) {
+      return a;
+    } else if (squares[a] && squares[a] === squares[c]) {
+      return b;
+    }
+  }
+
+  return null;
+}
+
+function calculateDraw(squares) {
+  for (let i = 0; i < squares.length; i++) {
+    if (!squares[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function calculateWinner(squares) {
+  for (let i = 0; i < lines.length; i++) {
+    const [a, b, c] = lines[i];
+    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
+      const winnerLine = lines[i];
+      const winnerMark = squares[a];
+      return [winnerMark, winnerLine];
+    }
+  }
+  return null;
+}
+
+function getRandomInt(min, max){
+  min = Math.ceil(min);
+  max = Math.floor(max);
+  return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+function getRandomAnglePosition(squares){
+  let squareNumber = getRandomInt(0,8);
+  while((squareNumber % 2 !== 0) || (squareNumber === 4) || squares[squareNumber]) {
+    squareNumber = getRandomInt(0,8);
+  }
+  return squareNumber;
+}
+
+function getRandomSidePosition(squares){
+  let squareNumber = getRandomInt(0,8);
+  while((squareNumber % 2 !== 1) || (squareNumber === 4) || squares[squareNumber]) {
+    squareNumber = getRandomInt(0,8);
+  }
+  return squareNumber;
+}
```

Comment:
I guess `getRandomAnglePosition` and `getRandomSidePosition` differ only by one `parameter`, use can unit them in 1 method

Reply:
Thanks, I replace functions "getRandomAnglePosition", "getRandomSidePosition" and "getRandomPosition" with:
```
function getRandomPosition(squares, where){
  let squareNumber = getRandomInt();

  if (where) {
    const remainder = where === 'onAngle' ? 0 : 1;

    while((squareNumber % 2 !== remainder) || (squareNumber === 4) || squares[squareNumber]) {
      squareNumber = getRandomInt();
    }
  } else {
    while(squares[squareNumber]) {
      squareNumber = getRandomInt();
    }
  }

  return squareNumber;
}
```